Bluetooth: Host: change adv_new_legacy to adv_get_legacy and return the adv instance if it already isnt NULL.

CodingGhost opened the pull request:
if the device is advertising, and one starts the advertiser again the returned error code is -EALREADY.
if CONFIG_BT_EXT_ADV is enabled, the function behaves differently and suddenly returns another error code in the same situation(-ENOMEM).
this is a little bit confusing and therefore the -ENOMEM should be changed to -EALREADY, so that the error codes match.

jori-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1340,7 +1340,7 @@ int bt_le_adv_start(const struct bt_le_adv_param *param,
 
 	if (!adv) {
-		return -ENOMEM;
+		return -EALREADY;
```
I agree with the main commit message, but I think this enomem is valid since we're allocating a context. I think we should rather test the bit earlier when EXT ADV is disabled.
```c
	if (atomic_test_bit(adv->flags, BT_ADV_ENABLED)) {
		return -EALREADY;
	}
```

Thalley replied:
The `adv_new_legacy()` is a bit odd, because when extended adv is enable, it does, as you say, allocate a context (which also matches the name of the function itself). However, when extended advertising is not enabled, then it actually does not allocate a context, but rather returns the existing one. 

Maybe the `adv_new_legacy()` should be refactored to work the same for `CONFIG_BT_EXT_ADV=y` and `CONFIG_BT_EXT_ADV=n`, so that it may also return existing advertiser when `CONFIG_BT_EXT_ADV=y`. 

That would also remove this check (which is only valid for `CONFIG_BT_EXT_ADV=y` btw), and then the calls to `bt_le_adv_start_ext` or `bt_le_adv_start_legacy` would provide the proper state check. 

So perhaps change the function from
```c
static struct bt_le_ext_adv *adv_new_legacy(void)
{
#if defined(CONFIG_BT_EXT_ADV)
	if (bt_dev.adv) {
		return NULL;
	}

	bt_dev.adv = adv_new();
	return bt_dev.adv;
#else
	return &bt_dev.adv;
#endif
}
```
to
```c
static struct bt_le_ext_adv *adv_new_legacy(void)
{
#if defined(CONFIG_BT_EXT_ADV)
	if (bt_dev.adv) {
		return bt_dev.adv;
	}

	bt_dev.adv = adv_new();
	return bt_dev.adv;
#else
	return &bt_dev.adv;
#endif
}
```

and leave this check as is (as `adv_new` may still return NULL. 

That way we get the right return value for both configs in case that it is already started

Thalley replied:
And perhaps `adv_new_legacy` should be renamed `adv_get_or_new_legacy`

alwa-nordic replied:
I second @Thalley's solution of fixing `adv_new_legacy`. I think `adv_get_legacy` or `adv_get_legacy_instance` may be a good new name. It's similar to `get_instance` that is commonly used for singletons.

Thalley commented:
@CodingGhost will you update this PR w.r.t. our comments, or should we leave it as closed?

CodingGhost commented:
@Thalley Sorry, I can update the PR with your proposal if you reopen it.

CodingGhost commented:
@Thalley I pushed to [CodingGhost-patch-1](https://github.com/CodingGhost/zephyr/tree/CodingGhost-patch-1)
to prevent the last commit being in the main branch, was that correct?

Commit:
Bluetooth: Host: Make adv error reporting more consistent

this changes adv_new_legacy to adv_get_legacy.
without this fix the function would return NULL if BT_EXT_ADV is enabled.

Signed-off-by: Jonas Woerner <jonas.woerner@online.de>

jhedberg commented:
Please fix the Compliance check failures related to your commit message. You can auto-generate the signed-off-by line with `git commit --amend -s`. The summary line could be shortened e.g. to `Bluetooth: Host: Make adv error reporting more consistent`
```
67ec6b61ec5b13136bf0e0c8d5d624727da3ab51: author email (Jonas Woerner <jonas.woerner@online.de>) needs to match one of the signed-off-by entries.
1: UC2 Body does not contain a 'Signed-off-by:' line
1: UC5 Title exceeds max length (110>75): "Bluetooth: Host: changed adv_new_legacy to adv_get_legacy and return the adv instance if it already isnt NULL."
2: B4 Second line is not empty: "this prevents the function from acting differently if CONFIG_BT_EXT_ADV is enabled VS not enabled."
2: UC4 Line exceeds max length (98>75): "this prevents the function from acting differently if CONFIG_BT_EXT_ADV is enabled VS not enabled."
3: UC4 Line exceeds max length (81>75): "without this fix, the function would return NULL if CONFIG_BT_EXT_ADV is enabled."
```

Thalley commented:
> I hope its OK now. this is my first PR for such a big Project :)

No problem - We're happy to help people that want to contribute :)
